### PR TITLE
Executor 2.0: Per-operator stream assignment policy

### PR DESCRIPTION
**Pull request by mzient**
## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
The PR implements a per-operator stream assignment algorithm.
The algorithm minimizes the number of necessary streams and tries to minimize the number of stream switches.

### Tests:

- [ ] Existing tests apply
- [X] New tests added
  - [ ] Python tests
  - [X] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [X] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [X] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/executor/executor2/stream_assignment.h [hunk 1]
```diff
@@ -164,6 +162,199 @@ class StreamAssignment<StreamPolicy::PerBackend> {
...
+    bool needs_stream = false;     // whether this exact node needs a stream
+    bool gpu_contributor = false;  // whether the node contributes to a GPU operator
+    std::optional<int> stream_id;
+    std::set<int> ready_streams;
+  };
+
+  void Assign(ExecGraph &graph) {
+    int num_nodes = graph.Nodes().size();
+
+    // the nodes in the graph must be sorted topologically
```
Comment (szkarpinski) [inline dali/pipeline/executor/executor2/stream_assignment.h:212]: Such requirements that are only expressed in the comments make this bug-prone. How about adding a simple check that the graph is actually toposorted? It's just something like `for each E in edges assert(e.from.index < e.to.index)` - linear time, few lines of code

File: dali/pipeline/executor/executor2/stream_assignment.h [hunk 2]
```diff
@@ -164,6 +162,199 @@ class StreamAssignment<StreamPolicy::PerBackend> {
...
+      ProcessNode(sorted_nodes_[i].first, sorted_nodes_[i].second);
+    }
+  }
+
+  void ProcessNode(const ExecNode *node, NodeMeta *meta) {
+    graph::Visit v(node);
+    if (!v)
+      return;
+
+    /* The algorithm
```
Comment (szkarpinski) [inline dali/pipeline/executor/executor2/stream_assignment.h:273]: I think we should add a note on complexity here, it's non-obvious for me, especially with the recursion in step 3

**Review by szkarpinski — APPROVED** [review]

File: dali/pipeline/executor/executor2/stream_assignment.h (hunk 1 above)
Comment (mzient) [reply dali/pipeline/executor/executor2/stream_assignment.h:212]: I should rephrase: the graph _is_ sorted. That's how it's built. You can manually construct one that's not, but it's not usable and its validation will fail.

File: dali/pipeline/executor/executor2/stream_assignment.h (hunk 2 above)
Comment (mzient) [reply dali/pipeline/executor/executor2/stream_assignment.h:273]: Outdated.

Comment (szkarpinski) [reply dali/pipeline/executor/executor2/stream_assignment.h:273]: Outdated

**Review by stiepan — APPROVED** [review]

**Pull request merged**